testing/sev_snp_hello_world_kernel/src/main.rs

```diff
@@ -16,22 +16,51 @@
 
 #![no_std]
 #![no_main]
+#![feature(abi_x86_interrupt)]
 
-use core::panic::PanicInfo;
+use core::{arch::asm, ops::Deref, panic::PanicInfo};
+use lazy_static::lazy_static;
 use oak_linux_boot_params::{BootParams, CCBlobSevInfo, CCSetupData, SetupDataType};
 use sev_guest::{
     cpuid::CpuidPage,
     msr::{get_sev_status, SevStatus},
     secrets::SecretsPage,
 };
-use x86_64::instructions::{hlt, interrupts::int3};
+use x86_64::{
+    instructions::{hlt, interrupts::int3},
+    structures::idt::{InterruptDescriptorTable, InterruptStackFrame},
+};
 
 mod asm;
 mod ghcb;
 mod serial;
 
+lazy_static! {
+    static ref IDT: InterruptDescriptorTable = {
+        let mut idt = InterruptDescriptorTable::new();
+        idt.general_protection_fault.set_handler_fn(gp_handler);
+        idt
+    };
+}
+
+extern "x86-interrupt" fn gp_handler(mut frame: InterruptStackFrame, _: u64) {
+    unsafe {
+        // RDMSR can cause a #GP if you try to read a non-existing MSR
+        if *frame.deref().instruction_pointer.as_ptr::<u16>() == 0x320Fu16 {
+            // just skip the instruction
+            frame.as_mut().update(|val| {
```

Comment:
Do you think it is perhaps also worth zeroing EAX and EDX to make sure it looks like the MSR returned 0? If we move the code around in future could it cuase stray values in the registers that could make the code think SEV-ES or SNP is enabled?

Reply:
Yes, you're correct, we're not properly emulating the `rdmsr` instruction but rather relying on the side effects here.

However, for a proper implementation we need to drop into raw assembly; you can't just set the register in Rust code. The `x86-interrupt` calling convention pushes RAX and RDX onto stack, and restores them from the stack before calling `iretq`. Therefore, even if we use `asm!` to write a value to RAX, it'll get overwritten. I don't know if there are any hooks for us to say that we _want_ to clobber RAX and RDX.

Reply:
I've added a comment stating that this handler doesn't emulate a `rdmsr` instruction, but rather just skips it.

Let's come back to it later, as we're likely going to need a better implementation for the Restricted Kernel anyway.